lib/resty/socket.lua

```diff
@@ -22,6 +24,16 @@ do
   end
 
   proxy_mt = {
+    connect = function(self, ...)
+      if self._connected_before then
+        -- luasocket does not allow reusing sockets after being closed
+        local socket = require 'socket'
+        self.sock = socket.tcp()
+      else
+        self._connected_before = true
+      end
+      return self.sock:connect(...)
+    end,
```

Comment:
Thinking about it, how about this, this should reduce the error margin:
```suggestion
    connect = function(self, ...)
      local ok, err = self.sock:connect(...)
      
      if self._connected_before and err == "closed" then
        -- luasocket does not allow reusing sockets after being closed
        local socket = require 'socket'
        self.sock = socket.tcp()
        ok, err = self.sock:connect(...)
      end
      
      self._connected_before = true
      return ok, err
    end,
```